### PR TITLE
Adding automatic mixed precision in create_supervised_evaluator

### LINKED ISSUE #1670: add automatic mixed precision in create_supervised_evaluator
## 🚀 Feature

Like #1235, add automatic mixed precision (both torch.cuda.amp) in `create_supervised_evaluator`.
PR that can reference: #1589 
Related issue: #1666

### PR DESCRIPTION
**Pull request by Moh-Yakoub**
Fixes #1670

Description:

Here I've refactored the evaluator and added support for `amp` mode for the `create_supervised_evaluator`. 
Since we don't calculate loss functions in evaluator, it doesn't make sense to add `apex` and `tpu` as separate steps as their equivalent trainers here in the PR 1589

I've submitted this to gather some feedback, once my approach is approved, I will update the docs and add tests in the same PR.

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: ignite/engine/__init__.py [hunk 1]
```diff
@@ -455,16 +455,134 @@ def create_supervised_evaluator(
...
+        The internal use of `device` has changed.
+        `device` will now *only* be used to move the input data to the correct device.
+        The `model` should be moved by the user before creating an optimizer.
+
+        For more information see:
+
+        - `PyTorch Documentation <https://pytorch.org/docs/stable/optim.html#constructing-it>`_
+
+        - `PyTorch's Explanation <https://github.com/pytorch/pytorch/issues/7844#issuecomment-503713840>`_
+    """
```
Comment (sdesrozis) [inline ignite/engine/__init__.py:572]: Add version changed in doc
```
    .. versionchanged:: 0.5.0
        - Added ``amp_mode`` argument for automatic mixed precision.
```

File: ignite/engine/__init__.py [hunk 2]
```diff
@@ -455,16 +455,134 @@ def create_supervised_evaluator(
...
+
+        The internal use of `device` has changed.
+        `device` will now *only* be used to move the input data to the correct device.
+        The `model` should be moved by the user before creating an optimizer.
+
+        For more information see:
+
+        - `PyTorch Documentation <https://pytorch.org/docs/stable/optim.html#constructing-it>`_
+
+        - `PyTorch's Explanation <https://github.com/pytorch/pytorch/issues/7844#issuecomment-503713840>`_
```
Comment (sdesrozis) [inline ignite/engine/__init__.py:508]: Add version changed in doc
```
    .. versionchanged:: 0.5.0
```

**Comment by sdesrozis** [comment]
@Moh-Yakoub Thank you. It looks good, I left minor comments.

**Comment by sdesrozis** [comment]
@Moh-Yakoub Thank you. 

@vfdev-5 @ydcjeff any comments ? Otherwise I merge the PR.

File: ignite/engine/__init__.py [hunk 3]
```diff
@@ -454,17 +454,143 @@ def create_supervised_evaluator(
...
+        device: device type specification (default: None).
+            Applies to batches after starting the engine. Model *will not* be moved.
+        non_blocking: if True and this copy is between CPU and GPU, the copy may occur asynchronously
+            with respect to the host. For other cases, this argument has no effect.
+        prepare_batch: function that receives `batch`, `device`, `non_blocking` and outputs
+            tuple of tensors `(batch_x, batch_y)`.
+        output_transform: function that receives 'x', 'y', 'y_pred' and returns value
+            to be assigned to engine's state.output after each iteration. Default is returning `(y_pred, y,)` which fits
+            output expected by metrics. If you change it you should use `output_transform` in metrics.
+        amp_mode (str, optional): can be ``amp`` or ``apex``, model and optimizer will be casted to float16 using
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:554]: There is no support for `apex` mode right ?

Comment (Moh-Yakoub) [reply ignite/engine/__init__.py]: updated the docs.

**Review by ydcjeff — COMMENTED** [review]
@Moh-Yakoub Thank you for the PR. A few comments and please update the docs too (in engine.rst).
Currently `supervised_evaluation_*` are not included in https://deploy-preview-1714--pytorch-ignite-preview.netlify.app

**Comment by ydcjeff** [comment]
@vfdev-5 @sdesrozis Since evaluation only has `torch.cuda.amp` to use for `amp`, can we merge `supervised_evaluation_*` in `create_supervised_evaluator` ? What do you think ?

File: tests/ignite/engine/test_create_supervised.py [hunk 4]
```diff
@@ -259,6 +262,18 @@ def test_create_supervised_evaluator_on_cuda_with_model_on_cpu():
     _test_create_supervised_evaluator(evaluator_device="cuda")
 
 
+@pytest.mark.skipif(LooseVersion(torch.__version__) < LooseVersion("1.6.0"), reason="Skip if < 1.6.0")
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="Skip if no GPU")
+def test_create_supervised_evaluator_on_cuda_amp():
+    model_device = trainer_device = "cuda"
```
Comment (ydcjeff) [inline tests/ignite/engine/test_create_supervised.py:268]:
```suggestion
    model_device = evaluator_device = "cuda"
```

File: tests/ignite/engine/test_create_supervised.py [hunk 5]
```diff
@@ -259,6 +262,18 @@ def test_create_supervised_evaluator_on_cuda_with_model_on_cpu():
...
+    _test_create_supervised_evaluator(model_device=model_device, amp_mode="amp")
```
Comment (ydcjeff) [inline tests/ignite/engine/test_create_supervised.py:269]:
```suggestion
    _test_create_supervised_evaluator(model_device=model_device, evaluator_device=evaluator_device, amp_mode="amp")
```

**Comment by vfdev-5** [comment]
> @vfdev-5 @sdesrozis Since evaluation only has `torch.cuda.amp` to use for `amp`, can we merge `supervised_evaluation_*` in `create_supervised_evaluator` ? What do you think ?

Do you mean to use the same `supervised_evaluation_step_amp` with enabled=True / False ?


Btw, thanks a lot for the review !

**Comment by ydcjeff** [comment]
> Do you mean to use the same supervised_evaluation_step_amp with enabled=True / False ?

Yes, we can even put them in `create_supervised_evaluator`

**Comment by vfdev-5** [comment]
> > Do you mean to use the same supervised_evaluation_step_amp with enabled=True / False ?
> 
> Yes, we can even put them in `create_supervised_evaluator`

Yes, it makes sense. @sdesrozis what do you think ?

**Comment by sdesrozis** [comment]
> > > Do you mean to use the same supervised_evaluation_step_amp with enabled=True / False ?
> > 
> > 
> > Yes, we can even put them in `create_supervised_evaluator`
> 
> Yes, it makes sense. @sdesrozis what do you think ?

The idea was to provide reusable codes. I agree that it was more relevant for training rather than evaluation since there are only 2 cases. Let's see first when docs will be fixed.

**Comment by Moh-Yakoub** [comment]
> > Do you mean to use the same supervised_evaluation_step_amp with enabled=True / False ?
> 
> Yes, we can even put them in `create_supervised_evaluator`

I don't mind either ways since it's a single function. even thought I always lean towards having a separate method to be reusable in the future.

File: tests/ignite/engine/test_create_supervised.py [hunk 6]
```diff
@@ -93,7 +96,7 @@ def _test_create_supervised_evaluator(
         example_input = torch.randn(1, 1)
         model = torch.jit.trace(model, example_input)
 
-    evaluator = create_supervised_evaluator(model, device=evaluator_device)
+    evaluator = create_supervised_evaluator(model, device=evaluator_device, amp_mode=amp_mode)
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:99]: Could you please check that `evaluator._process_function` is `supervised_evaluation_step` or `supervised_evaluation_step_amp`

**Review by ydcjeff — APPROVED** [review]
LGTM Thanks @Moh-Yakoub

File: tests/ignite/engine/test_create_supervised.py (hunk 6 above)
Comment (vfdev-5) [reply tests/ignite/engine/test_create_supervised.py:99]: @Moh-Yakoub could you please the above check here and we can merge this PR ?

Comment (Moh-Yakoub) [reply tests/ignite/engine/test_create_supervised.py:99]: @vfdev-5 sure, adding it now.

File: tests/ignite/engine/test_create_supervised.py [hunk 7]
```diff
@@ -93,7 +102,9 @@ def _test_create_supervised_evaluator(
...
+
+    assert evaluator._process_function is supervised_evaluation_step or supervised_evaluation_step_amp
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:107]: Well, the point here is to assert the function depending on `amp_mode` as it is always one of those 2 functions.

Comment (Moh-Yakoub) [reply tests/ignite/engine/test_create_supervised.py:107]: ah ok my bad. changing now.

Comment (Moh-Yakoub) [reply tests/ignite/engine/test_create_supervised.py]: @vfdev-5  also I feel that it's more standard to use `assert_called` instead of using the `process_function`. 
I feel that using `process_function` defies the point of the test as it requires us to know how the implementation is. while the unittest should just test the function behavior. what do you think?

Comment (vfdev-5) [reply tests/ignite/engine/test_create_supervised.py]: Yes, that's true and asserting if it was called is also a correct wat to do that but requires execution. On the other hand, here we test the function that sets up an engine and to go fast we just assert that it did what we'd like. I think it is OK for now here.

Comment (vfdev-5) [reply tests/ignite/engine/test_create_supervised.py]: According the the CI failure, we do not test it as we intend to. Failure is correct that process function is not the function used to create the wrapper... Probably, there is no way to check it ...

Comment (Moh-Yakoub) [reply tests/ignite/engine/test_create_supervised.py]: Yes I guess that the best way for now is to remove that assertion. I am not sure testing that the method was called is possible without mocking after execution

Comment (vfdev-5) [reply tests/ignite/engine/test_create_supervised.py]: Sounds good! Let's think about how to mock things in other issues/PRs. If you'd like to make a progress on that, let's expose the ideas in an issue and discuss there about what would be interesting to do.
